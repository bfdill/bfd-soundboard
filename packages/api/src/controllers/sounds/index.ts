import * as KoaRouter from 'koa-router'
import { NOT_FOUND, OK } from 'http-status-codes'
import { ISoundService, SoundService } from '@talkboy-9000/sounds'

import { IApplicationContext, ApplicationState } from '../../types'

export interface ISoundsController {
  get: (ctx: IApplicationContext) => void
}

export class SoundsController implements ISoundsController {
  private static instance: ISoundsController | undefined

  constructor(protected readonly soundService: ISoundService) {}

  get = async (context: IApplicationContext) => {
    const logger = context.logger.child({
      controller: {
        name: 'SoundsController',
        method: 'get'
      }
    })
    const sounds = this.soundService.getSounds(logger)

    context.body = { sounds }

    if (sounds === null) {
      logger.warn('no sounds from service')
      context.status = NOT_FOUND
      return
    }

    logger.info({
      soundsLength: sounds.length
    })
    context.status = OK
  }

  static getInstance(): ISoundsController {
    if (this.instance !== undefined) return this.instance

    this.instance = new SoundsController(SoundService.getInstance())

    return this.instance
  }
}

export const getSoundsRouter = () =>
  new KoaRouter<ApplicationState, IApplicationContext>().get(
    '/',
    SoundsController.getInstance().get
  )
