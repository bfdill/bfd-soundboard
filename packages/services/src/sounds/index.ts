import { join } from 'path'

export const PATH_TO_SOUNDS = join(process.cwd(), 'audio')
