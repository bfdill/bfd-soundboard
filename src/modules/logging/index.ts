import * as winston from 'winston'

export const transports = [
  new winston.transports.Console({
    level: 'silly',
    format: winston.format.prettyPrint()
  })
]

export const logger = winston.createLogger({
  transports,
  defaultMeta: {
    module: 'default'
  }
})

export const createModuleLogger = (module: string) =>
  winston.createLogger({
    transports,
    defaultMeta: { module },
    levels: winston.config.npm.levels
  })
