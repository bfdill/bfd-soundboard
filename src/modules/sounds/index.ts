import { join, resolve, sep, basename } from 'path'
import * as winston from 'winston'
import * as sane from 'sane'
import { sync } from 'glob'
import { getAppLogger } from '../../app'

export const PATH_TO_SOUNDS = join(process.cwd(), 'audio')

export type Sound = {
  id: string
  filename: string
  basename: string
}

export interface ISoundService {
  getBySoundId(soundId: string, parentLogger: winston.Logger): Sound | undefined
  getSounds: (parentLogger: winston.Logger) => Sound[]
  isPathValid: (filename: string, parentLogger: winston.Logger) => boolean
}

export class SoundService implements ISoundService {
  private static instance: ISoundService | undefined
  public readonly FILE_GLOB = '**/*.mp3'

  protected sounds: Sound[] = []
  protected readonly watcher: sane.Watcher

  constructor(
    public readonly pathToSounds: string,
    protected readonly saneFunction: typeof sane,
    protected readonly logger: winston.Logger
  ) {
    this.addSounds()
    this.watcher = this.createWatch()
  }

  addSounds = () => {
    this.logger.info('add sounds')

    sync(join(this.pathToSounds, this.FILE_GLOB)).forEach(this.addSound)
  }

  addSound = (filename: string) => {
    this.logger.debug({ filename })

    this.sounds = [
      ...this.sounds,
      {
        filename,
        id: filename,
        basename: basename(filename)
      }
    ]
      .sort((a, b) => a.filename.localeCompare(b.filename))
      .map((v, i) => {
        v.id = `${i}`
        return v
      })

    this.logger.silly({
      filename,
      sounds: this.sounds
    })
  }

  createWatch = (): sane.Watcher => {
    return this.saneFunction(this.pathToSounds, { glob: this.FILE_GLOB })
      .on('add', (path: string) => {
        this.addSound(path)

        this.logger.debug({
          path,
          message: 'watch:add'
        })
      })
      .on('delete', (path: string) => {
        this.removeSound(path)

        this.logger.debug({
          path,
          message: 'watch:delete'
        })
      })
  }

  getBySoundId = (
    soundId: string,
    parentLogger: winston.Logger
  ): Sound | undefined => {
    const result = this.sounds.find(s => s.id === soundId)
    parentLogger
      .child({
        service: {
          name: 'SoundService',
          method: 'getBySoundId'
        }
      })
      .debug({
        result,
        soundId,
        message: `getBySoundId(${soundId})`
      })
    return result
  }

  getSounds = (parentLogger: winston.Logger): Sound[] => {
    parentLogger
      .child({
        service: {
          name: 'SoundService',
          method: 'getSounds'
        }
      })
      .silly({ sounds: this.sounds })
    return this.sounds
  }

  isPathValid = (filename: string, parentLogger: winston.Logger): boolean => {
    const absCandidate = resolve(filename) + sep
    const result =
      absCandidate.substring(0, PATH_TO_SOUNDS.length) === PATH_TO_SOUNDS
    parentLogger
      .child({
        service: {
          name: 'SoundService',
          method: 'isPathValid'
        }
      })
      .debug({
        filename,
        result,
        message: `isPathValid(${filename})`,
        pathToSounds: PATH_TO_SOUNDS
      })
    return result
  }

  removeSound = (filename: string) => {
    this.logger.debug(`removeSound(${filename})`)

    const soundIndex = this.sounds.findIndex(s => s.filename === filename)

    if (soundIndex === -1) return

    this.sounds = this.sounds
      .slice(0, soundIndex)
      .concat(this.sounds.slice(soundIndex + 1))
      .sort((a, b) => `${a.filename}`.localeCompare(b.filename))
      .map((v, i) => {
        v.id = `${i}`
        return v
      })

    this.logger.silly({
      message: `removeSound(${filename})`,
      sounds: this.sounds
    })
  }

  static getInstance(): ISoundService {
    if (this.instance !== undefined) return this.instance

    this.instance = new SoundService(
      PATH_TO_SOUNDS,
      sane,
      getAppLogger().child({
        service: {
          name: 'SoundService'
        }
      })
    )

    return this.instance
  }
}
