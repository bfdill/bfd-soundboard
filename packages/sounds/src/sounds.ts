import * as sane from 'sane'
import * as winston from 'winston'
import { sync } from 'glob'
import { join, resolve, sep, basename } from 'path'
import { ISoundService, LOGGER_META } from './sounds.types'
import { Sound } from '@talkboy-9000/models'
import {
  Sounds as SoundConfig,
  ConfigurationService
} from '@talkboy-9000/configuration'

export class SoundService implements ISoundService {
  private static instance: ISoundService | undefined

  protected sounds: Sound[] = []
  protected readonly watcher: sane.Watcher

  constructor(
    public readonly config: SoundConfig,
    protected readonly saneFunction: typeof sane,
    protected readonly logger: winston.Logger
  ) {
    this.addSounds()
    this.watcher = this.createWatch()
  }

  addSounds = () => {
    this.logger.info('add sounds')

    sync(join(this.config.PathToSounds, this.config.FileGlob)).forEach(
      this.addSound
    )
  }

  addSound = (filename: string) => {
    this.logger.debug({ filename })

    this.sounds = [
      ...this.sounds,
      {
        filename,
        id: filename,
        basename: basename(filename)
      }
    ]
      .sort((a, b) => a.filename.localeCompare(b.filename))
      .map((v, i) => {
        v.id = `${i}`
        return v
      })

    this.logger.silly({
      filename,
      sounds: this.sounds
    })
  }

  private createWatch = (): sane.Watcher => {
    return this.saneFunction(this.config.PathToSounds, {
      glob: this.config.FileGlob
    })
      .on('add', (path: string) => {
        this.addSound(path)

        this.logger.debug({
          path,
          message: 'watch:add'
        })
      })
      .on('delete', (path: string) => {
        this.removeSound(path)

        this.logger.debug({
          path,
          message: 'watch:delete'
        })
      })
  }

  getBySoundId = (
    soundId: string,
    parentLogger: winston.Logger
  ): Sound | undefined => {
    const result = this.sounds.find(s => s.id === soundId)
    parentLogger
      .child({
        service: {
          name: 'SoundService',
          method: 'getBySoundId'
        }
      })
      .debug({
        result,
        soundId,
        message: `getBySoundId(${soundId})`
      })
    return result
  }

  getSounds = (parentLogger: winston.Logger): Sound[] => {
    parentLogger
      .child({
        service: {
          name: 'SoundService',
          method: 'getSounds'
        }
      })
      .silly({ sounds: this.sounds })
    return this.sounds
  }

  isPathValid = (filename: string, parentLogger: winston.Logger): boolean => {
    const absCandidate = resolve(filename) + sep
    const result =
      absCandidate.substring(0, this.config.PathToSounds.length) ===
      this.config.PathToSounds

    parentLogger
      .child({
        service: {
          name: 'SoundService',
          method: 'isPathValid'
        }
      })
      .debug({
        filename,
        result,
        message: `isPathValid(${filename})`,
        pathToSounds: this.config.PathToSounds
      })

    return result
  }

  removeSound = (filename: string) => {
    this.logger.debug(`removeSound(${filename})`)

    const soundIndex = this.sounds.findIndex(s => s.filename === filename)

    if (soundIndex === -1) return

    this.sounds = this.sounds
      .slice(0, soundIndex)
      .concat(this.sounds.slice(soundIndex + 1))
      .sort((a, b) => `${a.filename}`.localeCompare(b.filename))
      .map((v, i) => {
        v.id = `${i}`
        return v
      })

    this.logger.silly({
      message: `removeSound(${filename})`,
      sounds: this.sounds
    })
  }

  static getInstance(): ISoundService {
    if (this.instance !== undefined) return this.instance

    this.instance = new SoundService(
      ConfigurationService.getInstance().getSounds(),
      sane,
      winston.createLogger({ defaultMeta: LOGGER_META })
    )

    return this.instance
  }
}
