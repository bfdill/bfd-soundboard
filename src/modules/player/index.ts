// import * as playSound from 'play-sound'
import * as winston from 'winston'
import { ApplicationState } from '../../types'
import { createPlayer, IPlayer } from './types'
import { ISoundService, SoundService } from '../sounds'

// tsd creation fail
const createPlayer: createPlayer = require('play-sound')

export interface IPlayerService {
  playFile: (
    filename: string,
    state: ApplicationState,
    parentLogger: winston.Logger
  ) => PromiseLike<void>
}

export class PlayerService implements IPlayerService {
  static instance: IPlayerService | undefined

  constructor(
    protected readonly player: IPlayer,
    protected readonly soundService: ISoundService
  ) {}

  playFile = (
    filename: string,
    state: ApplicationState,
    parentLogger: winston.Logger
  ): PromiseLike<void> => {
    const logger = parentLogger.child({
      service: 'PlayerService',
      method: 'playFile'
    })

    logger.info({ filename })

    if (!this.soundService.isPathValid(filename, state, logger)) {
      const message = 'filename path is invalid'
      logger.error({ filename, message })
      return Promise.reject(message)
    }

    return new Promise((resolve, reject) => {
      const next = (error?: any) => {
        if (error === undefined || error === null) {
          logger.debug({
            filename,
            message: 'playback success'
          })
          resolve()
          return
        }

        logger.debug({
          error,
          filename,
          message: 'playback failure'
        })
        reject(error)
      }

      this.player.play(filename, next)
    })
  }

  static getInstance() {
    if (this.instance !== undefined) return this.instance

    this.instance = new PlayerService(
      createPlayer(),
      SoundService.getInstance()
    )

    return this.instance
  }
}
