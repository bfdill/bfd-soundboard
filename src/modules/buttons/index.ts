import * as winston from 'winston'
import { inspect } from 'util'
import * as rpiGpio from 'rpi-gpio'
import { ISoundService, SoundService } from '../sounds'
import { IPlayerService, PlayerService } from '../player'

export type Button = {
  switch: rpiGpio.Pins
  led: rpiGpio.Pins
  soundId: string
}

export type GpioSetupConfig = {
  pin: rpiGpio.Pins
  direction: rpiGpio.Direction
  edge: rpiGpio.Edge
  cb: () => void
}

export type SwitchState = {
  active: boolean
  pushed: boolean
  button: Button
}

const buttonList: Button[] = [
  { switch: '3', led: '8', soundId: '0' },
  { switch: '5', led: '10', soundId: '1' },
  { switch: '7', led: '12', soundId: '2' },
  { switch: '11', led: '16', soundId: '3' },
  { switch: '13', led: '18', soundId: 'rando' },
  { switch: '15', led: '22', soundId: 'rando' },
  { switch: '19', led: '24', soundId: '4' },
  { switch: '21', led: '26', soundId: '5' },
  { switch: '23', led: '32', soundId: '6' },
  { switch: '29', led: '36', soundId: '7' }
]

export interface IButtonService {
  ledOn: (button: Button, logger: winston.Logger) => Promise<void>
  setupFromConfig: (config: GpioSetupConfig) => Promise<void>
  registerChangeListener: (logger: winston.Logger) => () => void
  startButtonService: (baseLogger: winston.Logger) => Promise<void>
}

export class ButtonService implements IButtonService {
  private static instance: IButtonService | undefined
  protected readonly gpiop: rpiGpio.IGpioPromise
  protected readonly registeredSwitches: {
    [key: string]: SwitchState
  } = {}

  constructor(
    protected readonly rpiGpio: rpiGpio.IGpio,
    public readonly buttons: Button[],
    protected readonly soundService: ISoundService,
    protected readonly playerService: IPlayerService
  ) {
    this.gpiop = rpiGpio.promise
  }

  ledOn = (button: Button, logger: winston.Logger): Promise<void> => {
    logger.debug({
      button,
      method: 'ledOn'
    })

    return this.gpiop.write(button.led, true)
  }

  protected static setupCallback(
    button: Button,
    pin: rpiGpio.Pins,
    logger: winston.Logger
  ) {
    return (err: any) => {
      logger.log({
        button,
        pin,
        err,
        level: err ? 'error' : 'debug',
        message: 'setupCallback'
      })

      if (err) throw err
    }
  }

  protected static ledConfigFromButton(
    button: Button,
    logger: winston.Logger
  ): GpioSetupConfig {
    return {
      pin: button.led,
      direction: 'out',
      edge: 'none',
      cb: () => this.setupCallback(button, button.led, logger)
    }
  }

  protected static switchConfigFromButton(
    button: Button,
    logger: winston.Logger
  ): GpioSetupConfig {
    return {
      pin: button.switch,
      direction: 'in',
      edge: 'both',
      cb: () => this.setupCallback(button, button.switch, logger)
    }
  }

  setupFromConfig = (config: GpioSetupConfig) =>
    this.gpiop
      .setup(config.pin, config.direction, config.edge)
      .then(() => config.cb())

  registerChangeListener = (logger: winston.Logger) => () => {
    this.rpiGpio.on('change', (channel: rpiGpio.Pins, value: boolean) => {
      const sw = this.registeredSwitches[channel]
      const pushed = !value

      if (sw === undefined || sw === null) return
      if (sw.active === pushed) return
      if (!pushed) {
        sw.active = false
        return
      }

      sw.active = true
      const fileToPlay = this.soundService.getBySoundId(
        sw.button.soundId,
        logger
      )

      if (fileToPlay === undefined) {
        throw new Error(
          `unknown fileToPlay for button(${inspect(sw, false, 5)})`
        )
      }

      this.playerService.playFile(fileToPlay.filename, logger)
    })
  }

  startButtonService = (baseLogger: winston.Logger) => {
    const logger = baseLogger.child({
      service: 'ButtonService',
      method: 'startButtonService'
    })

    return Promise.all(
      this.buttons
        .map(btn => [
          this.setupFromConfig(ButtonService.ledConfigFromButton(btn, logger)),
          this.setupFromConfig(
            ButtonService.switchConfigFromButton(btn, logger)
          )
        ])
        .flat()
    )
      .then(() => {
        this.buttons.forEach(btn => {
          this.registeredSwitches[btn.switch] = {
            active: false,
            button: btn,
            pushed: false
          }
        })
      })
      .then(() => logger.info('setup ðŸ‘œðŸ‘œðŸ‘œ complete'))
      .then(() =>
        logger.debug({
          message: 'registeredSwitches',
          registeredSwitches: this.registeredSwitches
        })
      )
      .then(this.registerChangeListener(logger))
      .catch(cause => {
        logger.error({
          cause,
          message: 'startButtonService error'
        })
      })
  }

  static getInstance(): IButtonService {
    if (this.instance !== undefined) return this.instance

    return new ButtonService(
      rpiGpio,
      buttonList,
      SoundService.getInstance(),
      PlayerService.getInstance()
    )
  }
}
