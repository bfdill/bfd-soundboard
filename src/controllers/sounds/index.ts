import * as KoaRouter from 'koa-router'
import { NOT_FOUND, OK } from 'http-status-codes'

import { IApplicationContext, ApplicationState } from '../../types'
import { ISoundService, soundService } from '../../modules/sounds'

export interface ISoundsController {
  get: (ctx: IApplicationContext) => void
}

export class SoundsController implements ISoundsController {
  private static instance: ISoundsController | undefined

  constructor(protected readonly soundService: ISoundService) {}

  get = async (context: IApplicationContext) => {
    const logger = context.logger.child({
      controller: {
        name: 'SoundsController',
        method: 'get'
      }
    })
    const sounds = this.soundService.getSounds()

    context.body = { sounds }

    if (sounds === null) {
      logger.warn('no sounds from service')
      context.status = NOT_FOUND
      return
    }

    logger.info({
      soundsLength: sounds.length
    })
    context.status = OK
  }

  static getInstance(): ISoundsController {
    if (this.instance !== undefined) return this.instance

    this.instance = new SoundsController(soundService)

    return this.instance
  }
}

export const getSoundsRouter = () =>
  new KoaRouter<ApplicationState, IApplicationContext>().get(
    '/',
    SoundsController.getInstance().get
  )
