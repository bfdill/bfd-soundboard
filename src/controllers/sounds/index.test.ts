import { SoundsController, soundsControllerLogger } from '.'
import { ISoundService, Sound } from '../../modules/sounds'
import koaRouter = require('koa-router')

describe('controllers -> sounds -> index', () => {
  const mockSound: Sound = {
    basename: 'mock basename',
    filename: 'mock filename',
    id: 'mock id'
  }

  test.each([
    [null],
    [mockSound]
  ])('get matches snapshot', async (sounds?: Sound[] | null) => {
    const mockSoundService: ISoundService = {
      getSounds: () => sounds
    } as any
    const controller = new SoundsController(mockSoundService, soundsControllerLogger)
    const ctx: koaRouter.RouterContext = {} as any

    await controller.get(ctx)

    expect(ctx).toMatchSnapshot()
  })
})
