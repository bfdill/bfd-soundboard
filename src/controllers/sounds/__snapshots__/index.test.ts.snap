// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`controllers -> sounds -> index get matches snapshot 1`] = `
Object {
  "body": Object {
    "sounds": null,
  },
  "status": 404,
}
`;

exports[`controllers -> sounds -> index get matches snapshot 2`] = `
Object {
  "body": Object {
    "basename": "mock basename",
    "filename": "mock filename",
    "id": "mock id",
  },
}
`;
